### PR TITLE
DSD-1308: standalone link variant

### PR DESCRIPTION
**Pull request by bigfishdesign13**
Fixes JIRA ticket [DSD-1308](https://jira.nypl.org/browse/DSD-1308)

## This PR does the following:

- Adds the `"standalone"` variant to the `Link` component.
- Adds `NYPL Patterns` to the `Links Accessibility Guide`.

## How has this been tested?

- local Storybook

## Accessibility concerns or updates

- Style changes and recommended patterns have been updated based on UX exploration and input from a11y consultant.

### Checklist:

- [x] I have updated the Storybook documentation accordingly.
- [x] I have added relevant accessibility documentation for this pull request.
- [x] All new and existing tests passed.

### Front End Review:

- [x] Review the Vercel preview deployment once it is ready.

### CODE REVIEW

File: src/components/Link/Link.mdx [hunk 1]
```diff
@@ -36,9 +36,10 @@ import * as LinkStories from "./Link.stories";
 
 ## Accessibility
 
-The `Link` component should be used for navigation. If an `onClick` user
-interface action is required, a `Button` component should be used instead. The
-`Link` component renders an `<a>` element with the `href` attribute.
+The `Link` component renders an `<a>` element with the `href` attribute, so the
+`Link` component should be used for navigation. If an `onClick` user interface
+action is required, a `Button` component should be used instead. In simple
+terms, a link goes somewhere and a button does something.
```
Comment (EdwinGuzman) [inline src/components/Link/Link.mdx]: I think we should mention that it still is possible to pass an `onClick` to the link component because there are some very specific use cases and it is also a native HTML pattern that is allowed (but the HTML attribute is `onclick` all lowercase). We just really urge against it unless they know what they are doing.

File: src/components/Link/Link.tsx [hunk 2]
```diff
@@ -118,6 +122,30 @@ function getExternalExtraElements(
   );
 }
 
+function getStandaloneIcon(children: JSX.Element, linkId: string) {
+  const iconId = `${linkId}-icon`;
```
Comment (EdwinGuzman) [inline src/components/Link/Link.tsx]: The `Link`'s id prop is optional so this could turn into `"undefined-icon"`. We need a default or a better way to create this id.

File: src/components/Link/Link.tsx [hunk 3]
```diff
@@ -118,6 +122,30 @@ function getExternalExtraElements(
...
+  const extraElements = (
+    <>
```
Comment (EdwinGuzman) [inline src/components/Link/Link.tsx]: You don't need to wrap this in a fragment. Unless if it throws an error?

File: src/components/Link/Link.tsx [hunk 4]
```diff
@@ -118,6 +122,30 @@ function getExternalExtraElements(
...
+      <Icon
+        align={"right"}
+        className="more-link"
```
Comment (EdwinGuzman) [inline src/components/Link/Link.tsx]: Not a big deal but I think we use this only for tests, right? Maybe this should be `data-testid` for our unit tests.

File: src/components/Link/Link.tsx [hunk 5]
```diff
@@ -18,6 +18,7 @@ export const linkTypesArray = [
   "default",
   "external",
   "forwards",
+  "standalone",
```
Comment (EdwinGuzman) [inline src/components/Link/Link.tsx:22]: In the Link storybook doc, there's no example of this new variant... but also there are no real example of the other variants so this can also be updated in a separate PR.

File: src/components/AccessibilityGuide/Links.mdx [hunk 6]
```diff
@@ -50,7 +50,74 @@ format are generally considered to be fine without an underline.
...
+When a link is applied to a full heading, the link text SHOULD be rendered using
+the `link primary` color and the default state of the link text SHOULD NOT be
+underlined. The link text hover state SHOULD be rendered using the `link secondary` color and MUST be underlined. Applying a link to a full heading is
+the recommended pattern.
+
+When a link is applied to a subset of words within a heading, the link text MUST
+be rendered using the `link primary` color and the default state of the link
+text MUST be underlined. The link text hover state SHOULD be rendered using the
+`link secondary` color and MUST be underlined. While this is acceptable,
+applying a link to a potion of a heading is not the recommended pattern.
```
Comment (EdwinGuzman) [inline src/components/AccessibilityGuide/Links.mdx]: Can this be bold to highlight that it's not recommended?

File: src/components/AccessibilityGuide/Links.mdx [hunk 7]
```diff
@@ -50,7 +50,74 @@ format are generally considered to be fine without an underline.
...
+
+### Mixed Text
+
+When a body text link sits within a block of plain (unlinked) text &mdash; a
+paragraph, a single setence, or even just a few words &mdash; the link text
```
Comment (EdwinGuzman) [inline src/components/AccessibilityGuide/Links.mdx]: minor typo `setence`.

File: src/components/AccessibilityGuide/Links.mdx [hunk 8]
```diff
@@ -50,7 +50,74 @@ format are generally considered to be fine without an underline.
 
 ## NYPL Patterns
 
-TBD
+Text links can be presented in three general scenarios: in a heading, mixed
+within a block of plain text or as an isolated navigation element. In all
```
Comment (EdwinGuzman) [inline src/components/AccessibilityGuide/Links.mdx]: nitpick: comma after "text" and before "or".

File: src/components/AccessibilityGuide/Links.mdx [hunk 9]
```diff
@@ -50,7 +50,74 @@ format are generally considered to be fine without an underline.
...
+When a body text link sits within a block of plain (unlinked) text &mdash; a
+paragraph, a single setence, or even just a few words &mdash; the link text
+SHOULD be rendered using the `link primary` color and the default state of the
+link text MUST be underlined. The link text hover state SHOULD be rendered using
+the `link secondary` color and MUST be underlined.
+
+### Standalone Links
+
+When a body text link sits by itself within the page content area &mdash; a Read
+More link at the end of a short description, a View In Catelog link on a book
```
Comment (EdwinGuzman) [inline src/components/AccessibilityGuide/Links.mdx]: minor typo `Catelog`

File: src/components/AccessibilityGuide/Links.mdx [hunk 10]
```diff
@@ -50,7 +50,74 @@ format are generally considered to be fine without an underline.
...
+`link primary` color. The link text hover state SHOULD be rendered using
+the `link secondary` color and MUST be underlined.
+
+### Tips for Writing Standalone Link Text
+
+Avoid using a URL as your text link - URLs, especially long ones, may cause
+reading difficulty for both humans and screen readers.
+
+Keep link text as short as possible - Instead of "you can contact us,"" use "contact
+us.""
```
Comment (EdwinGuzman) [inline src/components/AccessibilityGuide/Links.mdx]: extra `"` at the end of both examples.

File: src/components/AccessibilityGuide/Links.mdx [hunk 11]
```diff
@@ -50,7 +50,74 @@ format are generally considered to be fine without an underline.
...
+
+Link at least one full word (two is better) - Avoid using a single letter or
```
Comment (EdwinGuzman) [inline src/components/AccessibilityGuide/Links.mdx]: I think these can be two sentences instead of having the hyphen.

**Review by EdwinGuzman — CHANGES_REQUESTED** [review]
Looks good overall but some visual examples would be nice.

File: src/components/Link/Link.tsx (hunk 4 above)
Comment (bigfishdesign13) [reply src/components/Link/Link.tsx]: I think that is leftover from the original `Link` component.  There is currently no mention of `more-link` anywhere outside of the `Link` files.  And even with that, it only seems necessary in the tests.  I will remove from the Story page, but for now I am going to leave it in the tests.

File: src/components/Link/Link.tsx (hunk 5 above)
Comment (bigfishdesign13) [reply src/components/Link/Link.tsx:22]: I added a new `All Link Types` section.

File: src/utils/utils.ts [hunk 12]
```diff
@@ -125,3 +125,9 @@ export const truncateText = (text: string, truncateTextLength: number = 60) => {
   const updatedText = text.substring(0, truncateTextLength - 1);
   return `${updatedText.substring(0, updatedText.lastIndexOf(" "))}...`;
 };
+
+/** Prepare a string for use in an ID or class attribute */
+export const sanitizeStringForAttribute = (str: string) => {
```
Comment (EdwinGuzman) [inline src/utils/utils.ts:130]: I'm surprised we don't already have this.

**Review by EdwinGuzman — APPROVED** [review]
Looks good!
